### PR TITLE
Handle runs of delimiters in filenames

### LINKED ISSUE #13991: Replace deprecated usage of `pkg_resources`
`setuptools` has deprecated their `pkg_resources` sub-package in favor of stdlib packages. See https://setuptools.pypa.io/en/latest/pkg_resources.html

We appear to use it on only a few places:

```shellsession
$ git grep pkg_resources
tests/unit/forklift/test_legacy.py:import pkg_resources
tests/unit/forklift/test_legacy.py:                project.name, pkg_resources.safe_name(project.name).lower()
warehouse/forklift/legacy.py:import pkg_resources
warehouse/forklift/legacy.py:    prefix = pkg_resources.safe_name(project.name).lower()
warehouse/forklift/legacy.py:    if not pkg_resources.safe_name(filename).lower().startswith(prefix):
warehouse/utils/readme.py:import pkg_resources
warehouse/utils/readme.py:    return pkg_resources.get_distribution("readme-renderer").version
```

I'm not certain where to best find a replacement for [`safe_name()`](https://setuptools.pypa.io/en/latest/pkg_resources.html#parsing-utilities:~:text=on%20other%20lines.-,safe_name(name),-Return%20a%20%E2%80%9Csafe) yet. We might have existing comparable functionality in warehouse already.
**Feel free to omit this from the refactor if it's too complex**.

A similar refactor took place in `readme-renderer`, see https://github.com/pypa/readme_renderer/pull/281 for details.

---

**Good First Issue**: This issue is good for first time contributors. If you've already contributed to Warehouse, work on [another issue without this label](https://github.com/pypi/warehouse/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+-label%3A%22good+first+issue%22) instead. If there is not a corresponding pull request for this issue, it is up for grabs. For directions for getting set up, see our [Getting Started Guide](https://warehouse.pypa.io/development/getting-started/).

If you are working on this issue and have questions, feel free to ask them here, in the [`#pypa-dev` chat channel on Libera](https://web.libera.chat/#pypa-dev), the [PyPA Discord](https://discord.gg/pypa) or on the [Discourse](https://discuss.python.org/c/packaging/14).

**Screenshot Required**: *If your pull request makes a visual change*, include a screenshot of your update. This helps our team give you feedback faster.

### PR DESCRIPTION
**Pull request by di**
Fixes the issue raised here: https://github.com/pypi/warehouse/pull/14155#issuecomment-1644195831

Fixes #13991.

### CODE REVIEW

**Comment by di** [comment]
(This also removes one of our usages of `pkg_resources.safe_name`, towards #13991)

**Comment by takluyver** [comment]
I think this probably works, but I don't know precisely what rules packages generating wheels (principally setuptools) have been using for the filename, e.g. if they might have `..` or `._` in the filename.

The [name normalization](https://packaging.python.org/en/latest/specifications/name-normalization/) spec has a one line function to fully normalize a name from any valid starting point. Is there any reason to do it differently for the filename prefix?

```python
def normalize(name):
    return re.sub(r"[-_.]+", "-", name).lower()
```

**Comment by di** [comment]
We don't actually _want_ to normalize the filename prefix, as it should already be normalized. But to handle backwards compatibility with tools that might not be replacing `-` and `.` with `_`, or lowercasing the prefix, we have to do a minimal normalization here, which we'll eventually want to remove via #14156:

```python
filename_prefix = filename_prefix.lower().replace(".", "_").replace("-", "_")
```

If we remove consecutive runs when normalizing the prefix, this would mean any filenames that have any amount of consecutive runs of delimiters in the filename would become valid, because the comparison between the prefix and the project name would not take repeating delimiters into account. Since these filenames are currently invalid, we haven't had an upload with a `._`, `_.`, or `..` in the filename for quite some time:

```
warehouse=> select upload_time, filename from release_files where filename ilike '%\.\_%.whl' order by upload_time desc limit 1;
        upload_time         |                   filename
----------------------------+-----------------------------------------------
 2021-11-03 20:54:45.340789 | hydrotools._restclient-3.0.5-py3-none-any.whl
(1 row)

warehouse=> select upload_time, filename from release_files where filename ilike '%\_\.%.whl' order by upload_time desc limit 1;
        upload_time         |                     filename
----------------------------+---------------------------------------------------
 2021-05-03 00:44:10.904378 | cmc_csci046_.yilinli_trees-1.0.1-py3-none-any.whl
(1 row)

warehouse=> select upload_time, filename from release_files where filename ilike '%\.\.%.whl' order by upload_time desc limit 1;
        upload_time         |                       filename
----------------------------+-------------------------------------------------------
 2020-11-06 20:00:06.720522 | example_..._pkg_megankuooo1234-0.0.1-py3-none-any.whl
(1 row)
```

So TL;DR: I'm not sure we want to become less restrictive by normalizing out character runs in the filename before comparison to the project name.

**Comment by takluyver** [comment]
Thanks, that makes sense. In any case, my main interest in the original
issue was for pre-normalised names, which should work now.

File: tests/unit/forklift/test_legacy.py [hunk 1]
```diff
@@ -2791,16 +2789,29 @@ def storage_service_store(path, file_path, *, meta):
...
     ):
         user = UserFactory.create()
         EmailFactory.create(user=user)
-        project = ProjectFactory.create(name="flufl.enum")
+        project = ProjectFactory.create(name=project_name)
         RoleFactory.create(user=user, project=project)
 
-        filename = "flufl_enum-1.0.0-py3-none-any.whl"
-        filebody = _get_whl_testdata(name="flufl_enum", version="1.0.0")
+        filename = filename_prefix + "-1.0.0-py3-none-any.whl"
```
Comment (miketheman) [inline tests/unit/forklift/test_legacy.py:2813]: f-strings!
```suggestion
        filename = f"{filename_prefix}-1.0.0-py3-none-any.whl"
```

**Review by miketheman — APPROVED** [review]
Good to go. Accept the suggestion if you like, it's okay if you don't!
